Math/NumberTheory/Moduli/CubicSymbol.hs

```diff
@@ -0,0 +1,100 @@
+module Math.NumberTheory.Moduli.CubicSymbol
+    ( CubicSymbol(..)
+    , conj
+    , cubicSymbol
+    ) where
+
+import Math.NumberTheory.Quadratic.EisensteinIntegers
+import qualified Math.NumberTheory.Quadratic.EisensteinIntegers as E
+
+import qualified Math.NumberTheory.Primes as P
+import qualified Data.Euclidean as A
+import Data.Semigroup as S
+
+
+
+data CubicSymbol = Zero | Omega | OmegaSquare | One deriving (Eq)
+
+instance S.Semigroup CubicSymbol where
+    (<>) = multiplication
+
+instance Show CubicSymbol where
+    show x
+        | x == Zero         = show 0
+        | x == Omega        = "ω"
+        | x == OmegaSquare  = "ω²"
+        | x == One          = show 1
+
+
+multiplication :: CubicSymbol -> CubicSymbol -> CubicSymbol
+multiplication x y
+    | x == Zero || y == Zero                   = Zero
+    | x == One                                 = y
+    | y == One                                 = x
+    | x == Omega && y == Omega                 = OmegaSquare
+    | x == Omega && y == OmegaSquare           = One
+    | x == OmegaSquare && y == Omega           = One
+    | x == OmegaSquare && y == OmegaSquare     = Omega
+    | otherwise                                = error ""
+
+exponentiation :: CubicSymbol -> Integer -> CubicSymbol
+exponentiation symbol exponent
+    | symbol == Zero || symbol == One          = symbol
+    | remainder == 0                           = One
+    | symbol == Omega && remainder == 1        = Omega
+    | symbol == Omega && remainder == 2        = OmegaSquare
+    | symbol == OmegaSquare && remainder == 1  = OmegaSquare
+    | symbol == OmegaSquare && remainder == 1  = Omega
+    | otherwise                                = error ""
+        where remainder = exponent `rem` 3
+
+conj :: CubicSymbol -> CubicSymbol
+conj x
+    | x == Zero || x == One      = x
+    | x == Omega                 = OmegaSquare
+    | x == OmegaSquare           = Omega
+    | otherwise                  = error ""
```

Comment:
Please use pattern matching instead of guards here and above.

Reply:
Done.